lib/evaluate/fold.cc

```diff
@@ -295,6 +295,72 @@ static std::optional<std::int64_t> GetInt64ArgOr(
   }
 }
 
+template<typename A, typename B>
+std::optional<std::vector<A>> GetIntegerVector(const B &x) {
+  static_assert(std::is_integral_v<A>);
+  if (const auto *someInteger{UnwrapExpr<Expr<SomeInteger>>(x)}) {
+    return std::visit(
+        [](const auto &typedExpr) -> std::optional<std::vector<A>> {
+          using T = ResultType<decltype(typedExpr)>;
+          if (const auto *constant{UnwrapConstantValue<T>(typedExpr)}) {
+            if (constant->Rank() == 1) {
+              std::vector<A> result;
+              for (const auto &value : constant->values()) {
+                result.push_back(static_cast<A>(value.ToInt64()));
+              }
+              return result;
+            }
+          }
+          return std::nullopt;
+        },
+        someInteger->u);
+  }
+  return std::nullopt;
+}
+
+template<typename T>
+std::optional<Constant<T>> Reshape(
+    FoldingContext &context, const ActualArguments &args) {
+  CHECK(args.size() == 4);
+  const auto *source{UnwrapConstantValue<T>(args[0])};
+  const auto *pad{UnwrapConstantValue<T>(args[2])};
+  std::optional<std::vector<ConstantSubscript>> shape{
+      GetIntegerVector<ConstantSubscript>(args[1])};
+  std::optional<std::vector<int>> order{GetIntegerVector<int>(args[3])};
+  if (source && shape && (pad || !args[2].has_value()) &&
+      (order || !args[3].has_value())) {
+    if (IsValidShape(shape.value())) {
+      int rank{GetRank(shape.value())};
+      if (const auto dimOrder{IsValidDimensionOrder(rank, order)}) {
+        std::size_t resultElements{TotalElementCount(shape.value())};
+        if (resultElements <= source->size() || (pad && !pad->empty())) {
+          Constant<T> result{!source->empty()
+                  ? source->Reshape(std::move(shape.value()))
+                  : pad->Reshape(std::move(shape.value()))};
+          ConstantSubscripts subscripts{result.lbounds()};
+          auto copied{result.CopyFrom(
+              *source, source->size(), subscripts, &dimOrder.value())};
+          if (copied < resultElements) {
+            CHECK(pad);
+            copied += result.CopyFrom(
+                *pad, resultElements - copied, subscripts, &dimOrder.value());
+          }
+          CHECK(copied == resultElements);
+          return result;
+        } else {
+          context.messages().Say(
```

Comment:
In cases where you emit warnings, you might want to return *some* result, so that multiple folds of the same expressions don't emit duplicate messages.

Reply:
It is a good point, however, I am not sure what that result should be... Should we have an `InvalidExpr<T>` to do this in a clean way ? The best I can think of right now is to change the intrinsic name to "invalid" while keeping the characteristic. I have applied it in this PR.

But what if the erroneous expression is nested into an inquiry function? e.g `RANK(RESHAPE(source, shape_with_too_many_dimensions)) ` Should this succeed returning size of `shape_with_too_many_dimensions` or should this fail ? With my change, it will succeed but emit a warning that RESHAPE is wrong.